src/main/java/org/manifold/compiler/back/digital/SMT2CodeGenerator.java

```diff
@@ -0,0 +1,299 @@
+import org.manifold.compiler.UndeclaredIdentifierException;
+import org.manifold.compiler.back.digital.smt2.Bitstring;
+import org.manifold.compiler.back.digital.smt2.QFABV;
+import org.manifold.compiler.back.digital.smt2.SExpression;
+import org.manifold.compiler.back.digital.smt2.Symbol;
+import org.manifold.compiler.middle.Schematic;
+
+public class SMT2CodeGenerator {
+  private static Logger log = LogManager.getLogger("SMT2CodeGenerator");
+
+  private Schematic schematic;
+  private Netlist netlist;
+  private PrimitiveTypeTable typeTable;
+  
+  private String outputDirectory;
+  public void setOutputDirectory(String dir) {
+    this.outputDirectory = dir;
+  }
+  
+  private boolean runChecks = true;
+  public void setRunChecks(boolean run) {
+    this.runChecks = run;
+  }
+  
+  // valid states are 0..numberOfStates inclusive
+  private int numberOfStates = 200;
+  public void setNumberOfStates(int n) {
+    this.numberOfStates = n;
+  }
+  
+  private List<SExpression> declarations;
+  private List<SExpression> assertions;
+  
+  public SMT2CodeGenerator(Schematic schematic, Netlist netlist,
+      PrimitiveTypeTable typeTable) {
+    this.schematic = schematic;
+    this.netlist = netlist;
+    this.typeTable = typeTable;
+    // by default, output to current working directory
+    this.outputDirectory = Paths.get("").toAbsolutePath().toString();
+    
+    this.declarations = new LinkedList<SExpression>();
+    this.assertions = new LinkedList<SExpression>();
+  }
+  
+  private void err(String message) {
+    throw new CodeGenerationError(message);
+  }
+  
+  private void generateInputPin(String nodeName, NodeValue node) 
+      throws UndeclaredIdentifierException {
+    // find out what net we drive
+    PortValue inputPort = node.getPort("out");
+    Net inputNet = netlist.getConnectedNet(inputPort);
+    String netName = inputNet.getName();
+    for (int i = 0; i <= numberOfStates; ++i) {
+      Symbol symInput = QFABV.getStateVariable(nodeName, i);
+      Symbol symNet = QFABV.getStateVariable(netName, i);
+      // declare state variables
+      declarations.add(QFABV.declareBitVector(symInput, 1));
+      declarations.add(QFABV.declareBitVector(symNet, 1));
+      // the input drives the net on every timestep
+      assertions.add(QFABV.assertThat(QFABV.equal(symInput, symNet)));
+    }
+  }
+  
+  private void generateOutputPin(String nodeName, NodeValue node) 
+      throws UndeclaredIdentifierException {
+    // find out what net drives the input
+    PortValue outputPort = node.getPort("in");
+    Net outputNet = netlist.getConnectedNet(outputPort);
+    String netName = outputNet.getName();
+    for (int i = 0; i <= numberOfStates; ++i) {
+      Symbol symOutput = QFABV.getStateVariable(nodeName, i);
+      Symbol symNet = QFABV.getStateVariable(netName, i);
+      // declare state variables (outputs only)
+      declarations.add(QFABV.declareBitVector(symOutput, 1));
```

Comment:
Should there be a ```declarations.add(QFABV.declareBitVector(symNet, 1));``` here similar to the generateInputPin function?

(also if that is the case, the for loop can be extracted to a common function between these 2)